### PR TITLE
[ROCm][Hardware][AMD] Use Triton Kernel for default FA on ROCm

### PR DESCRIPTION
**Pull request by jpvillam-amd**
This PR creates and makes default new triton exclusive backend for attention. Additionally removes some unsupported arguments from AMD's version of the `flash_attn_varlen_func` function.

- Changed selector to allow picking between multiple backends rather than just between two.
- Added a new triton FA backend available to ROCm. 
- Added new `VLLM_USE_FLASH_ATTN_TRITON` option to be able to swap between Triton and Default FA
- Removed unsupported attributes from AMD's `flash_attn_varlen_func`
- Added latest build of AMD's triton to Dockerfile.rocm

Appreciate any and all feedback, and apologies for the long PR

Co-authored-by: Vinayak Gokhale <vinayak.gokhale@amd.com>

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/attention/selector.py [hunk 1]
```diff
@@ -36,9 +36,12 @@ def _which_attn_to_use(dtype: torch.dtype) -> str:
...
     """
 
-    # NOTE: Defaulting to triton FA for AMD cards.
-    use_flash_attn_triton = os.environ.get('VLLM_USE_FLASH_ATTN_TRITON',
-                                           "True").lower() in ("true", "1")
+    # NOTE: Allow for switching between Triton and FA
+    #       Defaulting to triton FA for AMD cards.
+    use_flash_attn_triton = (os.environ.get("VLLM_USE_FLASH_ATTN_TRITON",
+                                            "True").lower()
+                             in ("true", "1")) and not is_hip()
```
Comment (hongxiayang) [inline vllm/attention/selector.py:43]: should it be and is_hip()?

Comment (jpvillam-amd) [reply vllm/attention/selector.py]: Correct mb.

**Review by hongxiayang — APPROVED** [review]
I have verified the changes on MI250x and MI300x. 
lgtm.

**Comment by hongxiayang** [comment]
Since latest ray has problem with tp>1 when running on AMD system now, can we also do this in `Dockerfile.rocm`:
`RUN python3 -m pip install --no-cache-dir ray[all]==2.9.3`

**Comment by gshtras** [comment]
> Since latest ray has problem with tp>1 when running on AMD system now, can we also do this in `Dockerfile.rocm`: `RUN python3 -m pip install --no-cache-dir ray[all]==2.9.3`

Also in requirements-rocm.txt

**Review by WoosukKwon — COMMENTED** [review]
Can we get rid of the CK-based FlashAttention, as we discussed offline in today's meeting?

File: vllm/attention/backends/flash_attn_triton.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+
+class FlashAttentionTritonBackend(FlashAttentionBackend):
+
+    @staticmethod
+    def get_impl_cls() -> Type["FlashAttentionTritonImpl"]:
+        return FlashAttentionTritonImpl
+
+
+class FlashAttentionTritonImpl(AttentionImpl):
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/attention/backends/flash_attn_triton.py:25]: Or, can we simply call this `ROCmFlashAttention`?

**Comment by hongxiayang** [comment]
> Can we get rid of the CK-based FlashAttention, as we discussed offline in today's meeting?

We can leave the clean up task in a different pull request.

File: vllm/attention/backends/flash_attn_triton.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jpvillam-amd) [reply vllm/attention/backends/flash_attn_triton.py:25]: Will do! I'll change it `TritonFlashAttention`, don't love `ROCmFlashAttention` as there isn't anything inherrently "ROCm" about it.

File: vllm/attention/backends/flash_attn.py [hunk 3]
```diff
@@ -192,19 +193,33 @@ def forward(
...
+                        cu_seqlens_q=attn_metadata.seq_start_loc,
+                        cu_seqlens_k=attn_metadata.seq_start_loc,
+                        max_seqlen_q=attn_metadata.max_prompt_len,
+                        max_seqlen_k=attn_metadata.max_prompt_len,
+                        softmax_scale=self.scale,
+                        causal=True,
+                        window_size=self.sliding_window,
+                        alibi_slopes=self.alibi_slopes,
+                    )
+
```
Comment (WoosukKwon) [inline vllm/attention/backends/flash_attn.py:222]: Actually in #3648, we are planning to use FlashAttention's recent APIs that support attention keys and values stored in paged KV cache. I believe the new APIs are incompatible with AMD GPUs.

Can we just use Triton FlashAttention at all times?

**Comment by WoosukKwon** [comment]
@jpvillam-amd Can we have `ROCmAttentionBackend` that selects one of the four different implementations of prefill Attention (i.e., Triton, CK, xFormers, naive) while using `PagedAttention` for decoding?

I feel like mixing the code for NVIDIA GPUs and AMD GPUs for attention is not good at the moment, since the APIs are not unified yet. For example, as I mentioned above, `FlashAttentionBackend` will use `flash-attn`'s paged KV cache APIs, which are not supported for AMD GPUs at the moment.

**Comment by jpvillam-amd** [comment]
> @jpvillam-amd Can we have `ROCmAttentionBackend` that selects one of the four different implementations of prefill Attention (i.e., Triton, CK, xFormers, naive) while using `PagedAttention` for decoding?
> 
> I feel like mixing the code for NVIDIA GPUs and AMD GPUs for attention is not good at the moment, since the APIs are not unified yet. For example, as I mentioned above, `FlashAttentionBackend` will use `flash-attn`'s paged KV cache APIs, which are not supported for AMD GPUs at the moment.

I will rename the backend and have it container all the selector logic for ROCm. I will leave xFormers out of the backend and have it selected with the current selector code same as CUDA 

@WoosukKwon Does that sounds ok?

**Comment by WoosukKwon** [comment]
@jpvillam-amd Thanks for updating the PR. Do you mind if I directly edit this PR? I refactored the PR a bit and wanted to directly upstream it to this PR for faster integration. If you don't mind, could you allow editing the PR? Your branch is protected now.

**Review by WoosukKwon — APPROVED** [review]
@jpvillam-amd LGTM. Thanks for updating the PR. Could you please submit another PR to test the backend? Thanks.

**Pull request merged**